internal/webhooks/patch_validation_test.go

```diff
@@ -0,0 +1,1372 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/internal/builder"
+)
+
+func TestValidatePatches(t *testing.T) {
+	tests := []struct {
+		name         string
+		clusterClass clusterv1.ClusterClass
+		wantErr      bool
+	}{
+		{
+			name: "pass multiple patches that are correctly formatted",
+			clusterClass: clusterv1.ClusterClass{
+				Spec: clusterv1.ClusterClassSpec{
+					ControlPlane: clusterv1.ControlPlaneClass{
+						LocalObjectTemplate: clusterv1.LocalObjectTemplate{
+							Ref: &corev1.ObjectReference{
+								APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+								Kind:       "ControlPlaneTemplate",
+							},
+						},
+					},
+
+					Patches: []clusterv1.ClusterClassPatch{
+						{
+							Name: "patch1",
+							Definitions: []clusterv1.PatchDefinition{
+								{
+									Selector: clusterv1.PatchSelector{
+										APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+										Kind:       "ControlPlaneTemplate",
+										MatchResources: clusterv1.PatchSelectorMatch{
+											ControlPlane: pointer.Bool(true),
+										},
+									},
+									JSONPatches: []clusterv1.JSONPatch{
+										{
+											Op:   "add",
+											Path: "/spec/template/spec/variableSetting/variableValue1",
+											ValueFrom: &clusterv1.JSONPatchValue{
+												Variable: pointer.String("variableName1"),
+											},
+										},
+									},
+								},
+							},
+						},
+						{
+							Name: "patch2",
+							Definitions: []clusterv1.PatchDefinition{
+								{
+									Selector: clusterv1.PatchSelector{
+										APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+										Kind:       "ControlPlaneTemplate",
+										MatchResources: clusterv1.PatchSelectorMatch{
+											ControlPlane: pointer.Bool(true),
+										},
+									},
+									JSONPatches: []clusterv1.JSONPatch{
+										{
+											Op:   "add",
+											Path: "/spec/template/spec/variableSetting/variableValue2",
+											ValueFrom: &clusterv1.JSONPatchValue{
+												Variable: pointer.String("variableName2"),
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					Variables: []clusterv1.ClusterClassVariable{
+						{
+							Name:     "variableName1",
+							Required: true,
+							Schema: clusterv1.VariableSchema{
+								OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+						{
+							Name:     "variableName2",
+							Required: true,
+							Schema: clusterv1.VariableSchema{
+								OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "pass if patches uses a name starting with \"builtin\" with no additional variable defined",
```

Comment:
nit, ~ oneOf:
```suggestion
			name: "pass if patch uses a variable which name starts with \"builtin\" with no additional variable defined",
```

```suggestion
			name: "pass if patches use a variable which name starts with \"builtin\" with no additional variable defined",
```

Reply:
Note: Absolutely up to you, but should this test be already here? Just seems to belong somewhere below to the variable test and I think it's ~ equivalent to `pass if jsonPatch uses a builtin variable which is defined`